### PR TITLE
Minor perf improvements to Sockets

### PR DESCRIPTION
**Pull request by stephentoub**
Shaves a few percent off various socket operations.  It's close to noise, but appears to be statistically significant and amount to around 2-3% improvement in microbenchmarks that repeatedly receive/send (in which case receives complete asynchronously) and send/receive (in which case most if not all operations complete synchronously).

(At this point most of the overhead associated with sockets (at least in the Windows build) is either in the P/Invokes to native or in peanut butter spread across lots of functions with very low exclusive amounts per function.  There's likely a decent amount more to be done on Linux, where we've done less profiling and where we know there are improvements to be made around asynchronous completion and dispatch.)

cc: @geoffkizer, @cipop, @davidsh, @vancem

### CODE REVIEW

File: src/Common/src/System/Net/SafeCloseSocket.Windows.cs [hunk 1]
```diff
@@ -33,52 +33,56 @@ public ThreadPoolBoundHandle IOCPBoundHandle
         // Binds the Socket Win32 Handle to the ThreadPool's CompletionPort.
         public ThreadPoolBoundHandle GetOrAllocateThreadPoolBoundHandle(bool trySkipCompletionPortOnSuccess)
         {
+            // Check to see if the socket native _handle is already
+            // bound to the ThreadPool's completion port.
+            if (_released || _iocpBoundHandle == null)
```
Comment (geoffkizer) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: Could we simplify this to a single test?  Is _iocpBoundHandle always null when _released == true?  Seems like it should be...

**Comment by geoffkizer** [comment]
Generally looks great; one small comment above.

File: src/Common/src/System/Net/SafeCloseSocket.Windows.cs (hunk 1 above)
Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: Maybe with other refactorings or changes, but at present it doesn't look like it... _iocpBoundHandle starts out as null and is then lazily initialized to non-null, after which point it's never set back to null.  From the comments in the code, it looks like this is done to support freeing a NativeOverlapped after disposal.

Comment (geoffkizer) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: Makes sense.

Another idea: is there any harm in just returning the existing iocp-bound handle if _released is true and _iocpBoundHandle is not null?

Don't know if any of this matters in practice, so don't block on this...

Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: You mean changing the condition from:
```C#
if (_released || _iocpBoundHandle == null) 
```
to instead just be:
```C#
if (_iocpBoundHandle == null) 
```
such that we don't throw the ObjectDisposedException from here but instead presumably have it throw some kind of exception when the handle is actually used?

Sounds reasonable, but I've not tried it.  Not sure if we have tests for that, either.  But it'd be worth checking on subsequently.

Comment (geoffkizer) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: Yep, that's what I was thinking.

**Pull request merged**